### PR TITLE
Refactor/collector

### LINKED ISSUE #5400: Failling test on master concerning metric,
```

--- jar coordinate com.fasterxml.jackson.core:jackson-databind already loaded with version 2.7.1 - omit version 2.7.1-1
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}
....................................................Updating dummy
Plugin foo version 9.0.0.0 does not exist
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................Validating dummy
ERROR: Configuration reloading also requires passing a configuration path with '-f yourlogstash.conf'
..............................F...........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Failures:

  1) LogStash::Pipeline when collecting metrics in the pipeline global metric populates the differents
     Failure/Error: expect(collected_metric[:stats][:events][:in].value).to eq(number_of_events)
       
       expected: 1000
            got: 1049
       
       (compared using ==)
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/logstash-core/spec/logstash/pipeline_spec.rb:618:in `(root)'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.8/lib/rspec/wait.rb:46:in `(root)'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/rakelib/test.rake:42:in `(root)'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/task.rb:248:in `execute'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/task.rb:243:in `execute'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/task.rb:187:in `invoke_with_call_chain'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/task.rb:180:in `invoke_with_call_chain'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/task.rb:173:in `invoke'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:150:in `invoke_task'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:106:in `top_level'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:106:in `top_level'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:115:in `run_with_threads'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:100:in `top_level'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:78:in `run'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:176:in `standard_exception_handling'
     # /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/vendor/bundle/jruby/1.9/gems/rake-11.1.2/lib/rake/application.rb:75:in `run'

Finished in 36.37 seconds (files took 3.48 seconds to load)
1330 examples, 1 failure

Failed examples:

rspec /var/lib/jenkins/workspace/elastic+logstash+master+multijob-os-compatibility/os/centos/logstash-core/spec/logstash/pipeline_spec.rb:617 # LogStash::Pipeline when collecting metrics in the pipeline global metric populates the differents

Randomized with seed 8085

### PR DESCRIPTION
**Pull request by ph**
Refactor the way the metric interact with the collector.
This remove the need of a singleton and make the agent the creator of a collector instance.

Fixes #5400 #5366

### CODE REVIEW

**Comment by ph** [comment]
cc @andrewvc

**Comment by ph** [comment]
looking at the failed test. *again*

**Comment by ph** [comment]
Fixed the build.

File: logstash-core/lib/logstash/agent.rb [hunk 1]
```diff
@@ -112,8 +113,13 @@ def uptime
...
   end
 
+  def stop_collecting_metric
+    @collector.stop
+    @periodic_pollers.stop
+  end
+
   def shutdown
-    stop_background_services
+    stop_collecting_metric
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: should be plural. s/metric/metrics/

File: logstash-core/lib/logstash/agent.rb [hunk 2]
```diff
@@ -131,7 +137,7 @@ def running_pipelines?
   private
   def start_webserver
     options = {:http_host => @http_host, :http_port => @http_port, :http_environment => @http_environment }
-    @webserver = LogStash::WebServer.new(@logger, options)
+    @webserver = LogStash::WebServer.new(@logger, self, options)
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: +1 for fewer singletons!

File: logstash-core/lib/logstash/agent.rb [hunk 3]
```diff
@@ -266,17 +266,15 @@ def running_pipeline?(pipeline_id)
   def upgrade_pipeline(pipeline_id, new_pipeline)
     stop_pipeline(pipeline_id)
     @pipelines[pipeline_id] = new_pipeline
+
+    new_pipeline = metric
```
Comment (andrewvc) [inline logstash-core/lib/logstash/agent.rb]: Why do we assign then not use this? Typo?

File: logstash-core/lib/logstash/api/rack_app.rb [hunk 4]
```diff
@@ -87,21 +88,23 @@ def self.app(logger, environment)
...
       end
 
-      def self.rack_namespaces
+      def self.rack_namespaces(agent)
         {
           "/_node" => LogStash::Api::Modules::Node,
           "/_stats" => LogStash::Api::Modules::Stats,
           "/_node/stats" => LogStash::Api::Modules::NodeStats,
-          "/_plugins" => LogStash::Api::Modules::Plugins
+          "/_plugins" => LogStash::Api::Modules::Plugins,
```
Comment (andrewvc) [inline logstash-core/lib/logstash/api/rack_app.rb]: You probably didn't mean to add this comma!

File: logstash-core/lib/logstash/api/service.rb [hunk 5]
```diff
@@ -58,15 +39,7 @@ def get(key)
...
-        (["LogStash::Instrument::MetricType::Counter", "LogStash::Instrument::MetricType::Gauge"] - metric_types).empty?
-      end
-
-      def metric_types
-        types = []
-        @snapshot_rotation_mutex.synchronize do
-          types = @snapshot.metric_store.all.map { |t| t.class.to_s }
-        end
-        return types
+        true
```
Comment (andrewvc) [inline logstash-core/lib/logstash/api/service.rb]: What is this useful for if it just returns true?

File: logstash-core/lib/logstash/pipeline.rb [hunk 6]
```diff
@@ -32,16 +32,16 @@ module LogStash; class Pipeline
...
+    :metric
 
   MAX_INFLIGHT_WARN_THRESHOLD = 10_000
 
   RELOAD_INCOMPATIBLE_PLUGINS = [
     "LogStash::Inputs::Stdin"
   ]
 
-  def initialize(config_str, settings = LogStash::SETTINGS)
+  def initialize(config_str, settings = LogStash::SETTINGS, provided_metric = nil)
```
Comment (andrewvc) [inline logstash-core/lib/logstash/pipeline.rb]: Could use a better name than `provided_metric`. Maybe `namespaced_metric` ?

File: logstash-core/spec/api/spec_helper.rb [hunk 7]
```diff
@@ -129,3 +86,20 @@ def self.included(klass)
     end
   end
 end
+
+shared_context "api setup" do
```
Comment (andrewvc) [inline logstash-core/spec/api/spec_helper.rb]: Love this refactor! 😎

File: logstash-core/spec/logstash/agent_spec.rb [hunk 8]
```diff
@@ -371,25 +384,32 @@
       subject.shutdown
       Stud.stop!(@t)
       @t.join
+      Thread.abort_on_exception = @abort_on_exception
```
Comment (andrewvc) [inline logstash-core/spec/logstash/agent_spec.rb]: Should probably be in an ensure block.

File: logstash-core/lib/logstash/api/service.rb (hunk 5 above)
Comment (ph) [inline logstash-core/lib/logstash/api/service.rb]: good catch, went I did the red green dance, I've tried to reduce the variables :)

**Comment by ph** [comment]
Updated with your code review.

**Comment by andrewvc** [comment]
LGTM!

**Comment by ph** [comment]
@purbon Can you also take a look?

**Comment by purbon** [comment]
LGTM, good work here.

**Pull request closed without merging**